include/ygm/container/oversized_bag.hpp

```diff
@@ -0,0 +1,335 @@
+      size_t target_rank;
+
+      // Determine target rank to match partitioning in ygm::container::array
+      if (idx < (global_size % m_comm.size()) * large_block_size) {
+        target_rank = idx / large_block_size;
+      } else {
+        target_rank = (global_size % m_comm.size()) +
+                      (idx - (global_size % m_comm.size()) * large_block_size) /
+                          small_block_size;
+      }
+
+      if (target_rank != m_comm.rank()) {
+        to_send[target_rank]++;
+      }
+    }
+    m_comm.barrier();
+
+    // Build and send oversized_bag indexes as calculated by to_send
+    for (auto &kv_pair : to_send) {
+      async_insert(local_pop(kv_pair.second), kv_pair.first);
+    }
+
+    m_comm.barrier();
+    */
+  }
+
+  /** @todo */  
+  template <typename RandomFunc>
+  void local_shuffle(RandomFunc &r) {
+    /*---This is the original code---
+    m_comm.barrier();
+    std::shuffle(m_local_bag.begin(), m_local_bag.end(), r);
+    */
+  }
+
+  /** @todo */
+  void local_shuffle() {
+    /*---This is the original code---
+    ygm::default_random_engine<> r(m_comm, std::random_device()());
+    local_shuffle(r);
+    */
+  }
+
+  /** @todo */
+  template <typename RandomFunc>
+  void global_shuffle(RandomFunc &r) {
+    /*---This is the original code---
+    m_comm.barrier();
+    std::vector<value_type> old_local_bag;
+    std::swap(old_local_bag, m_local_bag);
+
+    auto send_item = [](auto oversized_bag, const value_type &item) {
+      oversized_bag->m_local_bag.push_back(item);
+    };
+
+    std::uniform_int_distribution<> distrib(0, m_comm.size() - 1);
+    for (value_type i : old_local_bag) {
+      m_comm.async(distrib(r), send_item, pthis, i);
+    }
+    */
+  }
+
+  /** @todo */
+  void global_shuffle() {
+    /*---This is the original code---
+    ygm::default_random_engine<> r(m_comm, std::random_device()());
+    global_shuffle(r);
+    */
+  }
+
+  detail::round_robin_partitioner partitioner;
+
+ private:
+   /** @todo */
+  std::vector<value_type> local_pop(int n) {
+    /*---This is the original code---
+    YGM_ASSERT_RELEASE(n <= local_size());
+
+    size_t                  new_size  = local_size() - n;
+    auto                    pop_start = m_local_bag.begin() + new_size;
+    std::vector<value_type> ret;
+    ret.assign(pop_start, m_local_bag.end());
+    m_local_bag.resize(new_size);
+    return ret;
+    */
+  }
+
+  /** @todo */
+  void local_swap(self_type &other) { /* ---This is the original code--- m_local_bag.swap(other.m_local_bag); */ }
+
+  ygm::comm                       &m_comm;
+  /** @todo the default bag uses an std::vector as storage, we'll need a few things to do the oversized version. 
+   * My initial thoughts will be the following:
+   * 1) std::fstream object to write to a file
+   * 2) size_t local_size to keep track of the number of collective items in the local rank's files
+   * 3) size_t cur_size to keep track of the number of items in the current file
```

Comment:
so cur_size vs local_size, cur_size keeps track of the number of items that a rank is writing to/reading from, while local_size keeps track of how many items there are in total for that rank. Correct?

Reply:
Come back to this for the next PR, a lot of these implementation ideas changed.